juju/framework.py

```diff
@@ -253,6 +256,16 @@ def events(self):
                 events_map[event_kind] = bound_event
         return events_map
 
+    def __getitem__(self, key):
+        prefix = f'{key.replace("-", "_")}_'
+        scoped_events = {}
+        for event_name, bound_event in self.events().items():
+            if event_name.startswith(prefix):
+                unprefixed_event_name = event_name[len(prefix):]
+                scoped_events[unprefixed_event_name] = bound_event
+        events_scope = namedtuple('EventsScope', scoped_events.keys())
+        return events_scope(**scoped_events)
```

Comment:
That feels pretty heavy to have on every single attribute access. I think we can do that more cheaply and easily dynamically, by having an auxiliary object that would look a bit like this:
```
def __getitem__(self, key):
    return PrefixedEvents(self, key)
```
Makes sense?

Reply:
More specifically, we don't need any persistent storage of these attributes out of the real instance itself. PrefixedEvents would just look up the attribute on the real value on __getattr__ access.